### PR TITLE
2.x: Fix issues with MySQL database backup

### PR DESCRIPTION
**Pull request by alexislefebvre**
See #415

### CODE REVIEW

**Comment by alexislefebvre** [comment]
Some tests fail, if I run them separately then they pass. I'm wondering how could this happen.

**Comment by alexislefebvre** [comment]
I found a quick and dirty workaround:  Don't use backup service when appending data, see https://github.com/liip/LiipFunctionalTestBundle/pull/476/commits/dec6f27d05798745e03566e3fb0d5be154078ff4

**Comment by alexislefebvre** [comment]
I think that the backup service should keep an history of loaded fixtures so that if we load fixtures `A` then load fixtures `B` with `append = true`, the backup service recognizes that it can loads a backup with `A + B`. But I don't know if it's a good idea and if it's feasible.

File: src/Services/DatabaseBackup/AbstractDatabaseBackup.php [hunk 1]
```diff
@@ -33,18 +33,23 @@
...
     }
 
-    public function init(array $metadatas, array $classNames): void
+    public function init(array $metadatas, array $classNames, bool $append = false): void
     {
         $this->metadatas = $metadatas;
-        $this->classNames = $classNames;
+
+        if ($append) {
+            $this->classNames = array_merge($this->classNames, $classNames);
```
Comment (alexislefebvre) [inline src/Services/DatabaseBackup/AbstractDatabaseBackup.php]: When appending fixtures, keep track of `classNames` which had been already loaded.

File: src/Services/DatabaseBackup/MysqlDatabaseBackup.php [hunk 2]
```diff
@@ -84,7 +78,7 @@ public function backup(AbstractExecutor $executor): void
         $executor->getReferenceRepository()->save($this->getBackupFilePath());
         self::$metadata = $em->getMetadataFactory()->getLoadedMetadata();
 
-        exec("mysqldump -h $dbHost -u $dbUser -p$dbPass --no-create-info --skip-triggers --no-create-db --no-tablespaces --compact $dbName > {$this->getBackupFilePath()}");
+        exec("MYSQL_PWD=$dbPass mysqldump --host $dbHost --port=$dbPort --user $dbUser --no-create-info --skip-triggers --no-create-db --no-tablespaces --compact $dbName > {$this->getBackupFilePath()}");
```
Comment (alexislefebvre) [inline src/Services/DatabaseBackup/MysqlDatabaseBackup.php]: Using an environment variable remove this error: `Warning: Using a password on the command line interface can be insecure.`.

File: tests/AppConfigMysqlCacheDb/AppConfigMysqlKernelCacheDb.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * (c) Lukas Kahwe Smith <smith@pooteeweet.org>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Liip\FunctionalTestBundle\Tests\AppConfigMysqlCacheDb;
+
+/*
+ * This file is part of the Liip/FunctionalTestBundle
```
Comment (magnetik) [inline tests/AppConfigMysqlCacheDb/AppConfigMysqlKernelCacheDb.php]: Double header

**Comment by alexislefebvre** [comment]
@magnetik Yes we could add an annotation to disable the cache for a test.

**Comment by alexislefebvre** [comment]
@magnetik I created an issue with your idea: #480

**Comment by VincentClair** [comment]
Cache seems not be compatible with loadFixtureFiles().
The backup file is empty and $classNames argument of loadFixtures() empty too.
I have missconfigure something ?

**Comment by alexislefebvre** [comment]
If you upgraded the bundle, the old cache can cause issues. You can remove it and try again.

**Comment by VincentClair** [comment]
It's not a Symfony cache problem, I already drop completely the folder ;)

After investigation the code of the last alpha release, loadFixtureFiles() not used the cache system:
WebTestCase::loadFixtureFiles() calls AbstractDatabaseTool::loadAliceFixture(). Inside latter is called cleanDatabase() without parameters, that causing a empty cache file for database. Then fixtures are loaded from files thanks to fidry_alice_data_fixtures.loader.doctrine service.

Maybe there could be a cache for this external service from theofidry/alice-data-fixtures library to activate ?

**Pull request merged**